Bed leveling assessment popup

eduard-sukharev opened the pull request:
This feature adds a helpful popup window after bed leveling is complete to help users assess bed leveling results.

resolves #21

eduard-sukharev commented:
¡Hola, @navaismo!

Have a look at this PR - it has basic bed leveling assessment logic and should be ready for you to integrate it with your fancy bed leveling UI.
There's no actual Popup implementation yet (as you can tell adding a popup is quite mundane job), and the `HMI_Leveling_Assess_Popup` is more like a draft piece, not called from anywhere and should probably be removed, but shows how to call the bed level assessment implementation.

navaismo commented:
Hi @eduard-sukharev 
I pulled your branch and added the changes, but after commit and push VSCODE created a new branch xD:  https://github.com/navaismo/Ender-3V3-SE/tree/your_bed_level_is_ok

So I guess i couldn't puh it into your PR. well let me know if you want to download and work on the new branch, or maybe merge this on Main and I'll add the changes after merge.

eduard-sukharev commented:
Looks ok, left you a comment there, but it's minor.
I have ticked the Allow edits checkbox so that you can push to my branch (I suppose?)

![image](https://github.com/user-attachments/assets/e739cf63-8cc8-4bcb-9293-1026bb6c88f9)

This feature still needs polishing (the requested way to "disable" these obnoxiously helpful messages and popups so that advanced users don't receive them), but overall it's good to merge, I guess?

navaismo commented:
I guess I was able to push the changes into your PR/Branch, let me know.
So I can delete the branch on my repo right? and later when the work is done in the PR I can merge it.

eduard-sukharev commented:
¡Hola, Max!

I'm tinkering with the code you've provided and found that there's this `HMI_M117Info`, which looks like some hack or workaround for something. Is it safe to not use it or is it required for proper Octoprint integration? Can I remove the call to it and the `check_key` redirect to it?

navaismo commented:
привет!

`HMI_M117Info` Is used to go back the main Menu, it is used for the M117 Messages and for the BedLevel Visualizer, I guess for the Finish Job in Octoprint and maybe another one. It is generic so I re used it to avoid creating more HMI functions that do the same.

eduard-sukharev commented:
¡Hola, Max!

I have pushed my edits into this branch, but I'm currently facing an issue I cannot fix so far, and how no free time yet to tackle further.

The idea here is following - I have added an `ADVANCED_HELP_MESSAGES` config flag and there's also an additional `advanced_help_enabled_flag:1` settings bit flag (currently without it's own menu to be altered). If firmware is built with `ADVANCED_HELP_MESSAGES` then triggering `advanced_help_enabled_flag` in menu would replace the bed leveling screen with your 3d mesh screen, and actual mesh edit would be available directly from Control => Edit Bed Leveling.
The screen itself is great right now, with reworked thresholds and colors, but the issue I'm facing is that the main menu Bed Leveling action is handled very differently from other simple screens and menus. It draws empty window frame and then calls the G29 which in turn draws leveling grid items one by one and then leaves at that screen. My attempt to put 3d mesh screen there kinda worked, but it's only drawn once and not actually holds the screen logically - if you rotate the encoder it will update screen drawing Main screen icons on top of it. So, there seem to be an issue with how to "keep screen attention" on that simple "OK Dialog" that can only be quit with single encoder push.

I know, that all sounds pretty complex, but you'll see what I mean if you do leveling and then try to rotate encoder (or if you push encoder then it'll start leveling again).

If you have time, please have a look and try your fixes. This might be very nice to have and me stucking on it for too long won't do good to the project.

navaismo commented:
> If you have time, please have a look and try your fixes. This might be very nice to have and me stucking on it for too long won't do good to the project.

Sure I'll take a look on that next week and let you know if I have some doubts.

eduard-sukharev commented:
oops, haha, totally missed that. I hope you'll figure this out - you seem to be advancing at a very good pace lately!

navaismo commented:
Hi @eduard-sukharev  I just pushed some changes, removed the 'scroll' from encoder and just accepting the push of it in the HMI_OnlyConfirm, now is not showeing the main menu when the knob is moved right or left just the push is detected.

I like it a lot how it looks, beutiful work. Check if that works for you and maybe we can merge it on the current or for the next release.

eduard-sukharev commented:
I have resolved merge conflicts. Let me test it and see if we can roll it.

eduard-sukharev commented:
Ok, we need a menu item in Control to allow end users to toggle these messages. Let me finish it today.

eduard-sukharev commented:
Ok, I have replaced the "Show bed viz" menu item with the "Advanced help" toggle and made it to show as popup **before** auto bed level results, but not instead of them.
Have a look at it, test and see if it works ok with Octoprint. And if so, we can roll it :)

navaismo commented:
> found a slight issue that after bed leveling the popup with bed with shown, then actual bed grid values, and when user confirms UI shows the Control menu (as if user came to bed grid values from Edit bed leveling menu). Not a big deal

Happened to me, but actually I think was by design and for me is usefull to come back to the edit values again.  I will leave it like that.


> On a second thought, should we actually show the bed leveling grid if user opted for advanced help messages?

Yep I was going to ask where the menu went, If we nebaled that we should be able to see the Bed Level Visualizer menu is a good feature.

navaismo commented:
> Our menu system needs revamp, though

Yes, maybe as you once said create a menu for leveling stuff. Because redesign the main menu will be problematic to add new menus or layout due to the LCD limitations.

navaismo commented:
Pivet @eduard-sukharev 

Wanted to ask if you had issues testing the branch. Because I have some but not sure if related with the branch since we reached more than 50% of memory usage.

The issues I have right now are:

- Underextrusion, Clicking Extruder, super painful sound and the print has a lot of underextrusion on all layers or walls.
- Bed termistor failures, almost all my prints failed twice due to Low temperature detected, I see the values of termistor bouncing hard from 65 to 53 in one sec.
- Octoprint lose connection and commands became unresponsive, I Tried with the CacomixtlePad too and got the same issue.

Right now Im flashing, and configuring the old 1.0.9.8 which is under 50% to see if it is related or is just a coincidence and my printer is just dying. YA ne znayu, drug 😮‍💨 Maybe is time to install klipper, Im tired to print 5 things very good and suddenly everything falls apart.

So just wanted to ask to see if its time to throw my printer away.

eduard-sukharev commented:
If printing is what pays your bill then definitely go for the right tool for you. You don't owe anyone, do the best for you. I rarely use mine these days and this can afford keeping it in experimental state.

eduard-sukharev commented:
ok, let's do it. if have fix at hand — echa pa'palante!

eduard-sukharev commented:
If I get it right, QR codes only support uppercase characters in this mode and in this length, so I have added toupper inside the qr code generation code. I have now updated the bitly link, it should work now.
